### DISCUSSION TITLE
Sonoff TH16 Elite - missing display driver TM1621 LCD

### DISCUSSION THREAD

**Discussion by pmuller9999** [arendst/Tasmota#19745] [Q&A] [answered]
Hi,
I have with succes installed latest version of Tasoma ESP32-Display firmware on my Sonoff TH16 Elite.

I would like to be able to write to LCD-display. As I understand LCD chipset is TM1621. It also is consistent with configuration template of https://template.blakadder.com/sonoff_THR316D.html.

But I can't find driver available in binary eventhough I can find screenshots on Internet with that configuration. 

Do I have to add driver myself?

Br
Peter

**Comment by sfromis** [comment, marked as answer]
Do not use tasmota32-display, as that binary does not have the specialized driver for the combo of TM1621 and Sonoff display. Use the regular binary instead.

Then you can configure it, and get a default setup for what is displayed. You can also use the [DspLine](https://tasmota.github.io/docs/Commands/#dspline) command to select what sensors to include, examples in the [driver source code](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_87_esp32_sonoff_tm1621.ino).

**Reply by pmuller9999** [reply]
Thanks. That solved the problem. Also learned to use Partition Wizard;-)

**Reply by dr-knor** [reply]
@sfromis I have a related question. I want to use the TH16 Elite with multiple DB18S20 sensors and for that purpose need to add #define DS18x20_USE_ID_ALIAS in user_config_override.h and compile the binary myself. But I noticed that TM1621 is not supported in the configurations that I can select to compile (using Tasmocompiler).  Is there a way around this, e.g. making an addition to user_config_override.h to make it work. I tried #define USE_DISPLAY_TM1621_SONOFF but that results in a compilation error. With the regular tasmota32 binary the display is working OK, but then I cannot reliably use the multiple DB18S20 sensors.

**Reply by sfromis** [reply]
Hard to make guesses about "compilation error" without details.

Anyway, DS18B20 sensor usage is also reliable without aliasing, albeit less convenient, when you recognize sensors by their `Id` instead of the Tasmota-generated object name for the sensor. Of course, it should still be feasible to do a custom build with additional options.

**Reply by dr-knor** [reply]
Thanks! The error was not very descriptive (to me at least): *** [.pio/build/tasmota32/src/tasmota.ino.cpp.o] Error 1. And Exit code 1. But I will look into the option of using the Id for the sensors instead. It would anyway have my preference to use a custom build if possible.

**Reply by sfromis** [reply]
That's not the error message, just a termination message about failure. The actual error message is earlier in the log.

**Reply by dr-knor** [reply]
Sorry for my ignorance. I am getting there. Earlier on: /tmp/Tasmota/tasmota/tasmota_xdrv_driver/xdrv_87_esp32_sonoff_tm1621.ino:442:75: error: 'Energy' was not declared in this scope.
So I guess I have to include the feature that includes 'Energy'. I will have a look. I also looked at the option to use the Id but I get the impression that would mean using scripts instead of rules to make it work for my case.
Thanks again btw.

**Reply by dr-knor** [reply]
It's working. Compilation OK and I have a number on the display. Adding the 'Energy Sensors' feature did the trick. Thanks a million.

**Reply by sfromis** [reply]
Great :smile: 

While using the `Id` with rules would be a hassle, what I prefer would be neither rules nor the scripter, but instead the Berry language you get with ESP32 devices.
